internal/controllers/gateway/gateway_controller.go

```diff
@@ -0,0 +1,581 @@
+				})
+			}
+			if ingress.Hostname != "" {
+				addresses = append(addresses, gatewayv1alpha2.GatewayAddress{
+					Type:  &gatewayHostAddrType,
+					Value: ingress.Hostname,
+				})
+			}
+		}
+	}
+
+	return addresses, listeners, nil
+}
+
+// determineL6and7ListenersFromDataPlane takes a list of L4 Gateway listeners and references
+// them against the data-plane to determine any L6 (TLS) or L7 (HTTP, e.t.c.) support configured for them.
+func (r *GatewayReconciler) determineL6and7ListenersFromDataPlane(ctx context.Context, svc *corev1.Service, listeners []gatewayv1alpha2.Listener) ([]gatewayv1alpha2.Listener, error) {
+	// gather the proxy and stream listeners from the data-plane and map them
+	// to their respective ports (which will be the targetPorts of the proxy
+	// Service in Kubernetes).
+	proxyListeners, streamListeners, err := r.KongConfig.Client.Listeners(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve listeners from the data-plane: %w", err)
+	}
+	proxyListenersMap := make(map[int]kong.ProxyListener)
+	for _, listener := range proxyListeners {
+		proxyListenersMap[listener.Port] = listener
+	}
+	streamListenersMap := make(map[int]kong.StreamListener)
+	for _, listener := range streamListeners {
+		streamListenersMap[listener.Port] = listener
+	}
+
+	// map servicePorts to targetPorts in order to identify the data-plane
+	// proxy_listeners and stream_listeners for the service by their ports.
+	portMapper := make(map[int]int)
+	for _, port := range svc.Spec.Ports {
+		portMapper[int(port.Port)] = port.TargetPort.IntValue()
+	}
+
+	// upgrade existing L4 listeners with any higher level protocols that
+	// are configured for them in the data-plane.
+	upgradedListeners := make([]gatewayv1alpha2.Listener, 0, len(listeners))
+	for _, listener := range listeners {
+		if streamListener, ok := streamListenersMap[portMapper[int(listener.Port)]]; ok {
+			if streamListener.SSL {
+				listener.Protocol = gatewayv1alpha2.TLSProtocolType
+			}
+		}
+		if proxyListener, ok := proxyListenersMap[portMapper[int(listener.Port)]]; ok {
+			if proxyListener.SSL {
+				listener.Protocol = gatewayv1alpha2.HTTPSProtocolType
+			} else {
+				listener.Protocol = gatewayv1alpha2.HTTPProtocolType
+			}
+		}
+		upgradedListeners = append(upgradedListeners, listener)
+	}
+
+	return upgradedListeners, nil
+}
+
+// -----------------------------------------------------------------------------
+// Gateway Controller - Private Object Update Methods
+// -----------------------------------------------------------------------------
+
+// updateAddressesAndListenersSpec updates a unmanaged gateway's spec with new addresses and listeners.
+// If the addresses and listeners provided are the same as what exists, the update is skipped.
+func (r *GatewayReconciler) updateAddressesAndListenersSpec(ctx context.Context, gateway *gatewayv1alpha2.Gateway, addresses []gatewayv1alpha2.GatewayAddress, listeners []gatewayv1alpha2.Listener) (bool, error) {
```

Comment:
this logic is not unit-testable and I think it should be unit-tested

Reply:
The integration tests cover this logic, and the logic is bound up in the reconcilation loop pretty strongly. Can you help me to understand better what's missing and perhaps we can simply add some more integration tests to cover it?

Reply:
We have a chunk of logic that performs reflection magic. This is opaque and understandign that it does _exactly_ what it's intended to is IMO desirable. It's easy to lose this level of detail (specifically: that we use an arcane mechanism to do a simple thing) on the integration testing level.

What I'm looking for is a table-based unit test that has `gateway` `addresses`, `listeners` as inputs, and `true/false` and the resulting `gateway` as expected outputs.

Reply:
I'm having a hard time seeing `DeepEquals` as reflection magic. I think one could argue the function could be moved out of reflection because it's kind of a weird place for it to be, but that's neither here nor there. It's a very common tool used to compare the equality of nested structures. It sounds like we're asking to unit test `DeepEquals` which seems out of scope for this work.

Reply:
> It sounds like we're asking to unit test DeepEquals

We're asking to unit test a logical conjunction of two usages of reflection to perform comparison (which is surely unit tested, but could potentially be prone to corner cases) and a resulting decision.

The fact that we use an external library (reflect) instead of simple comparisons makes this even more reasonable to unit test IMO.

Reply:
To invert the question a little bit:

- Is the purpose of unit testing to detect corner case behavior and logic bugs?
- If the answer to the above question is yes: what differentiates this chunk of logic from another (hypothetical) chunk of logic that would deserve an unit test?

Reply:
We've always wanted to unit test everything, but we've settled for integration tests in the past. It sounds like this PR has become the duck in a game of duck duck goose for starting to re-apply pressure on that, which I suppose makes some sense anyway it's the first major change since 2.0. That said, I think we're missing the forest for the trees here: we need to be institutionalizing controller unit tests as opposed to picking out specific methods and intentionally wrapping them around sub-functions just so they can be partially testable. I've created https://github.com/Kong/kubernetes-ingress-controller/issues/2006 as a follow up to this, but I'll also do what is requested here and break this out into a unit-tested function in the meantime.

Reply:
After re-acquainting myself with `envtest` I remembered how heavyweight it is. I have a partial implementation in a branch now, but there's a _lot_ more work to do to get that done that I don't feel should block this. As such #2006 is the follow up, but I added unit tests for this in e9e436f18596b8c5f6945f31ed530a614fd3a860 as requested.